src/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareParameters.cs

```diff
@@ -0,0 +1,215 @@
+            if (customAttribute != null && customAttribute.Name != null)
+            {
+                mapping.Add(DTOproperty.Name, customAttribute.Name);
+            }
+        }
+        return mapping;
+    }
+}
+
+public class CarbonAwareParameters
+{
+    private IEnumerable<Location> _mulitpleLocations = new List<Location>();
+    private DateTimeOffset _start = DateTimeOffset.MinValue;
+    private DateTimeOffset _end = DateTimeOffset.MaxValue;
+
+    public enum PropertyName { MultipleLocations, Start, End };
+    public IEnumerable<Location> MultipleLocations {
+        get { return _mulitpleLocations; }
+        set { 
+            if (value.Any()) // Only set if non-empty
+            {
+                _mulitpleLocations = value; 
+                _props[PropertyName.MultipleLocations].IsSet = true;
+            }
+        }
+            
+    }
+    public DateTimeOffset Start {
+        get { return _start; }
+        set { _start = value; _props[PropertyName.Start].IsSet = true; }
+    }
+    public DateTimeOffset End {
+        get { return _end; }
+        set { _end = value; _props[PropertyName.End].IsSet = true; }
+    }
+
+    public class Property
+    {
+        public PropertyName Name { get; init; }
+        public bool IsSet = false;
+        public bool IsRequired = false;
+        public bool IsValid => (IsRequired && IsSet) || !IsRequired;
+        public string DisplayName { get; set; }
+
+        public Property(PropertyName name)
+        {
+            Name = name;
+            DisplayName = name.ToString();
+        }
+    }
+
+    internal Dictionary<PropertyName, Property> _props { get; init; }
+
+    public CarbonAwareParameters(Dictionary<string, string>? displayNameMap = null)
+    {
+        _props = InitProperties();
+        if (displayNameMap is not null) { _props.UpdateDisplayNames(displayNameMap); }
+    }
+
+    /// <summary>
+    /// Validates the properties and relationships between properties. This includes
+    ///  - Check that required properties are set
+    ///  - Check that specified relationships between properties (like start < end) are true
+    /// </summary>
+    public void Validate()
+    {
+        var errors = new Dictionary<string, List<string>>();
+
+        var multipleLocations = _props[PropertyName.MultipleLocations];
+        var start = _props[PropertyName.Start];
+        var end = _props[PropertyName.End];
+
+        // Validate Properties
+        if (!multipleLocations.IsValid) { errors.AppendValue(multipleLocations.DisplayName, $"{multipleLocations.DisplayName} is not set"); }
+        if (!start.IsValid) { errors.AppendValue(start.DisplayName, $"{start.DisplayName} is not set"); }
+        if (!end.IsValid) { errors.AppendValue(end.DisplayName, $"{end.DisplayName} is not set"); }
+
+        // Want to throw any property validation errors before checking any relationships between properties
```

Comment:
Perhaps rename to make it more explicit that the method throws an exception if there are errors?

Reply:
We would rather not do `ValidateOrThrow` since the assumption is generally that a validation function throws if any of the validations fail. We added better explanation in the summary and remarks to explain this behavior

Reply:
Agree with Validate(). My comment was for the naming of CheckErrors() method.